### ISSUE TITLE
Is it possible to automatically drain a backend in terminating status?

### ISSUE THREAD

**Issue by dcowden** [jcmoraisjr/haproxy-ingress#137]
We have a legacy application ( tomcat/java), which needs sticky sessions. When we deploy new versions of our applications, we need to stop sending new connections to a server, while sending bound sessions to the old server. Please note: this is not referring to in-flight requests, we're needing the active tomcat sessions to expire, which normally takes a few hours.

This is possible using haproxy drain command. This will send bound connections to the old server, but send new ones elsewhere. But in kubernetes, calling a command on the ingress controller is not part of the deployment flow. To do it with current tools, we would need to add a preStop hook to our application. In that hook, we'd need to access the ingress controller, and ask it to drain with an api call. We'd rather not introduce the ability for applications to call apis on the ingress controller.

When kubernetes terminates a pod, it enters the TERMINATING status. In nearly all cases, when sticky sessions are enabled, the desired functionality is probably to put the associated pod into drain mode. Is this possible with the nginx-plus ingress controller?

We currently use the kubernetes-maintained nginx ingress controller, but its drain command is a commercial only feature.  We've not had any reason to switch to using the haproxy ingress controller so far, but this would be one that would make it worth it.

How could we accomplish this flow using the haxproxy ingress controller?

**Comment by jcmoraisjr** [comment]
Perhaps you are talking about #95 ? This feature was just merged thanks to @brianloss . Please let me know if this is the feature you're looking for and give HAProxy Ingress a try.

**Comment by dcowden** [comment]
Wow SCORE!  yes, thats it! I'm sorry I didn't find that myself in a search. I'm going to try this out today.

**Comment by dcowden** [comment]
@jcmoraisjr what image version should i use to try this out? it looks like it was merged into master < 1 day ago, but the newest image on quay is v0.6-a80d2c6, which is about 14 days old.

**Comment by dcowden** [comment]
Re-opened, because I _think_ i dont have a way to try this out yet-- it was merged more recently than the last build.

**Comment by dcowden** [comment]
Hello @jcmoraisjr I tried this out today, using the latest image I could find ( quay.io/jcmoraisjr/haproxy-ingress:v0.6-a80d2c6 )

I am able to get the ingress controller running without issues, and sticky sessions work as expected. 

I also enabled the drain support as follows:

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: haproxy-ingress
data:
  balance-algorithm: "roundrobin"
  drain-support: "true"
  use-proxy-protcol: "true"
  ingress.kubernetes.io/ssl-redirect : "false"
```

When I delete a pod, it stays in terminating for a while  as expected ( since my pre-stop hook waits for sessions to die ), but haproxy-ingress removes the endpoint immediately, and sends traffic to a new node.

I suspect this is because #95 is not in this build. How can i get a later  build?

**Comment by dcowden** [comment]
Thank you. I will test this asap and report back.

**Comment by dcowden** [comment]
@jcmoraisjr I did a round of testing. Unfortunately, I found that with drain enabled, the haproxy instance crashed. 
Here are the logs: they don't seem all that helpful. I was unable to find out how to turn on debug, but i'll try it again:

```
Reload strategy: native
I0412 12:10:20.580628       9 launch.go:141] 
Name:       HAProxy
Release:    latest
Build:      git-2000ef7
Repository: https://github.com/jcmoraisjr/haproxy-ingress.git
I0412 12:10:20.580773       9 launch.go:144] Watching for ingress class: haproxy
I0412 12:10:20.581071       9 launch.go:340] Creating API client for https://10.96.0.1:443
I0412 12:10:20.590012       9 launch.go:352] Running in Kubernetes Cluster version v. (v1.9.4) - git (clean) commit bee2d1505c4fe820744d26d41ecd3fdd4a3d6546 - platform linux/amd64
I0412 12:10:20.591690       9 launch.go:168] validated default/ingress-default-backend as the default backend
I0412 12:10:20.595938       9 controller.go:1312] starting Ingress controller
I0412 12:10:20.603092       9 event.go:218] Event(v1.ObjectReference{Kind:"Ingress", Namespace:"default", Name:"testdrain", UID:"429b2be8-3daf-11e8-b00e-000c2999b7ca", APIVersion:"extensions", ResourceVersion:"173104", FieldPath:""}): type: 'Normal' reason: 'CREATE' Ingress default/testdrain
I0412 12:10:26.002378       9 controller.go:1320] running initial sync of secrets
I0412 12:10:26.002601       9 leaderelection.go:174] attempting to acquire leader lease...
I0412 12:10:26.008205       9 controller.go:311] backend reload required
I0412 12:10:26.011124       9 leaderelection.go:184] successfully acquired lease default/ingress-controller-leader-haproxy
I0412 12:10:26.011189       9 status.go:199] new leader elected: haproxy-ingress-lhgxs
I0412 12:10:26.025718       9 controller.go:320] ingress backend successfully reloaded...
I0412 12:11:39.826618       9 main.go:45] Shutting down with signal terminated
I0412 12:11:39.831712       9 controller.go:1298] shutting down controller queues
I0412 12:11:39.832763       9 status.go:123] updating status of Ingress rules (remove)
I0412 12:11:39.847188       9 status.go:142] removing address from ingress status ([])
I0412 12:11:39.849234       9 status.go:364] updating Ingress default/testdrain status to []
I0412 12:11:39.862607       9 main.go:38] Exiting (0)
```

**Comment by dcowden** [comment]
correction-- crashed is probably the wrong word. It exited with code zero, which means somehow it thought it needed to shut down.

**Comment by ausmith** [comment]
@dcowden I think you need to add `--v=3` according to https://github.com/jcmoraisjr/haproxy-ingress/issues/135#issuecomment-379941185

**Comment by dcowden** [comment]
ah ok @ausmith  thank you!

**Comment by dcowden** [comment]
@jcmoraisjr I've duplicated the problem, and attached the logs below.

My use case is:

- Start a script requesting urls from a load balanced set of 3 backends.
- Terminate one pod using kubectl delete po <podname>
- Observe the crash below.

Configuration:
```
apiVersion: extensions/v1beta1
kind: DaemonSet
metadata:
  labels:
    run: haproxy-ingress
  name: haproxy-ingress
spec:
  updateStrategy:
    type: RollingUpdate
  selector:
    matchLabels:
      run: haproxy-ingress
  template:
    metadata:
      labels:
        run: haproxy-ingress
    spec:
      hostNetwork: true
      serviceAccountName: ingress-controller
      containers:
      - name: haproxy-ingress
        image: quay.io/jcmoraisjr/haproxy-ingress:v0.6-2000ef7
        args:
        - --default-backend-service=$(POD_NAMESPACE)/ingress-default-backend
        - --allow-cross-namespace
        - --configmap=$(POD_NAMESPACE)/haproxy-ingress
        - --sort-backends
        - --v=3
        ports:
        - name: http
          containerPort: 80
        - name: https
          containerPort: 443
        - name: stat
          containerPort: 1936
        env:
        - name: POD_NAME
          valueFrom:
            fieldRef:
              fieldPath: metadata.name
        - name: POD_NAMESPACE
          valueFrom:
            fieldRef:
              fieldPath: metadata.namespace`
--
apiVersion: v1
kind: ConfigMap
metadata:
  name: haproxy-ingress
data:
  balance-algorithm: "roundrobin"
  drain-support: "true"
  use-proxy-protcol: "true"

```

Logs:

```I0412 16:19:59.522819      10 controller.go:1180] getting endpoints for service default/testdrain and port &ServicePort{Name:http,Protocol:TCP,Port:80,TargetPort:8080,NodePort:0,}
I0412 16:19:59.522831      10 controller.go:1248] endpoints found: [{192.168.224.6 8080 0 0 false &ObjectReference{Kind:Pod,Namespace:default,Name:testdrain-c5589c8bf-kvlll,UID:7f5a0cd8-3e5c-11e8-93f0-000c2999b7ca,APIVersion:,ResourceVersion:179394,FieldPath:,}} {192.168.224.7 8080 0 0 false &ObjectReference{Kind:Pod,Namespace:default,Name:testdrain-c5589c8bf-8rrx9,UID:cff18c8d-3e5e-11e8-93f0-000c2999b7ca,APIVersion:,ResourceVersion:180152,FieldPath:,}} {192.168.224.8 8080 0 0 false &ObjectReference{Kind:Pod,Namespace:default,Name:testdrain-c5589c8bf-ml87j,UID:c83189e4-3e50-11e8-93f0-000c2999b7ca,APIVersion:,ResourceVersion:175495,FieldPath:,}}]
I0412 16:19:59.522932      10 controller.go:1072] ingress default/testdrain for host draintest.colinx.com does not contains a TLS section
I0412 16:19:59.522940      10 main.go:73] Ingress testdrain: No value found in annotation ingress.kubernetes.io/session-cookie-name. Using the default INGRESSCOOKIE
I0412 16:19:59.522946      10 main.go:83] No annotation value found in Ingress testdrain: ingress.kubernetes.io/session-cookie-strategy. Setting it to default insert
I0412 16:19:59.522950      10 main.go:91] Invalid or no annotation value found in Ingress testdrain: ingress.kubernetes.io/session-cookie-hash. Setting it to default md5
I0412 16:19:59.523087      10 main.go:73] Ingress testdrain: No value found in annotation ingress.kubernetes.io/session-cookie-name. Using the default INGRESSCOOKIE
I0412 16:19:59.523092      10 main.go:83] No annotation value found in Ingress testdrain: ingress.kubernetes.io/session-cookie-strategy. Setting it to default insert
I0412 16:19:59.523095      10 main.go:91] Invalid or no annotation value found in Ingress testdrain: ingress.kubernetes.io/session-cookie-hash. Setting it to default md5
I0412 16:19:59.523513      10 controller.go:551] replacing ingress rule default/testdrain location / upstream default-testdrain-80 (upstream-default-backend)
I0412 16:19:59.523583      10 controller.go:331] obtaining information about stream services of type TCP located in configmap 
I0412 16:19:59.523590      10 controller.go:331] obtaining information about stream services of type UDP located in configmap 
I0412 16:19:59.523599      10 controller.go:307] skipping backend reload (no changes detected)
I0412 16:20:05.045215      10 queue.go:70] queuing item sync status
I0412 16:20:05.045420      10 queue.go:111] syncing sync status
I0412 16:20:05.051535      10 status.go:353] skipping update of Ingress default/testdrain (no change)
I0412 16:21:05.045881      10 queue.go:70] queuing item sync status
I0412 16:21:05.045970      10 queue.go:111] syncing sync status
I0412 16:21:05.060197      10 status.go:353] skipping update of Ingress default/testdrain (no change)
I0412 16:22:05.047320      10 queue.go:70] queuing item sync status
I0412 16:22:05.047390      10 queue.go:111] syncing sync status
I0412 16:22:05.058356      10 status.go:353] skipping update of Ingress default/testdrain (no change)
I0412 16:23:05.047699      10 queue.go:70] queuing item sync status
I0412 16:23:05.047736      10 queue.go:111] syncing sync status
I0412 16:23:05.051739      10 status.go:353] skipping update of Ingress default/testdrain (no change)
I0412 16:24:05.048773      10 queue.go:70] queuing item sync status
I0412 16:24:05.048818      10 queue.go:111] syncing sync status
I0412 16:24:05.052518      10 status.go:353] skipping update of Ingress default/testdrain (no change)
I0412 16:25:05.050171      10 queue.go:70] queuing item sync status
I0412 16:25:05.050280      10 queue.go:111] syncing sync status
I0412 16:25:05.068667      10 status.go:353] skipping update of Ingress default/testdrain (no change)
I0412 16:26:05.051390      10 queue.go:70] queuing item sync status
I0412 16:26:05.051453      10 queue.go:111] syncing sync status
I0412 16:26:05.056086      10 status.go:353] skipping update of Ingress default/testdrain (no change)
I0412 16:27:05.052089      10 queue.go:70] queuing item sync status
I0412 16:27:05.052148      10 queue.go:111] syncing sync status
I0412 16:27:05.056351      10 status.go:353] skipping update of Ingress default/testdrain (no change)
I0412 16:27:11.150966      10 queue.go:70] queuing item &Pod{ObjectMeta:k8s_io_apimachinery_pkg_apis_meta_v1.ObjectMeta{Name:testdrain-c5589c8bf-kvlll,GenerateName:testdrain-c5589c8bf-,Namespace:default,SelfLink:/api/v1/namespaces/default/pods/testdrain-c5589c8bf-kvlll,UID:7f5a0cd8-3e5c-11e8-93f0-000c2999b7ca,ResourceVersion:185112,Generation:0,CreationTimestamp:2018-04-12 14:19:22 +0000 UTC,DeletionTimestamp:2018-04-12 16:30:31 +0000 UTC,DeletionGracePeriodSeconds:*200,Labels:map[string]string{app: testdrain,pod-template-hash: 711457469,},Annotations:map[string]string{},OwnerReferences:[{extensions/v1beta1 ReplicaSet testdrain-c5589c8bf 9e6cecf1-3db5-11e8-b00e-000c2999b7ca 0xc4202d0e19 0xc4202d0e1a}],Finalizers:[],ClusterName:,Initializers:nil,},Spec:PodSpec{Volumes:[{colinx-token-gpcqw {nil nil nil nil nil SecretVolumeSource{SecretName:colinx-token-gpcqw,Items:[],DefaultMode:*420,Optional:nil,} nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil}}],Containers:[{testapp monitor-test [] []  [{ 0 6139 TCP } { 0 8080 TCP }] [] [{CATALINA_OPTS -Dcom.sun.management.jmxremote -Dcom.sun.management.jmxremote.port=6139 -Dcom.sun.management.jmxremote.ssl=false -Dcom.sun.management.jmxremote.authenticate=false nil}] {map[memory:{{1 9} {<nil>} 1G DecimalSI} cpu:{{1 0} {<nil>} 1 DecimalSI}] map[cpu:{{200 -3} {<nil>} 200m DecimalSI} memory:{{165675008 0} {<nil>} 158Mi BinarySI}]} [{colinx-token-gpcqw true /var/run/secrets/kubernetes.io/serviceaccount  <nil>}] Probe{Handler:Handler{Exec:nil,HTTPGet:&HTTPGetAction{Path:/monitor-test-webapp/api/heartbeat,Port:8080,Host:,Scheme:HTTP,HTTPHeaders:[],},TCPSocket:nil,},InitialDelaySeconds:10,TimeoutSeconds:20,PeriodSeconds:10,SuccessThreshold:1,FailureThreshold:3,} &Probe{Handler:Handler{Exec:nil,HTTPGet:&HTTPGetAction{Path:/monitor-test-webapp/api/heartbeat,Port:8080,Host:,Scheme:HTTP,HTTPHeaders:[],},TCPSocket:nil,},InitialDelaySeconds:10,TimeoutSeconds:20,PeriodSeconds:10,SuccessThreshold:1,FailureThreshold:3,} &Lifecycle{PostStart:nil,PreStop:&Handler{Exec:&ExecAction{Command:[sh /home/appuser/preStop.sh 120],},HTTPGet:nil,TCPSocket:nil,},} /dev/termination-log File Never nil false false false}],RestartPolicy:Always,TerminationGracePeriodSeconds:*30,ActiveDeadlineSeconds:nil,DNSPolicy:ClusterFirst,NodeSelector:map[string]string{},ServiceAccountName:colinx,DeprecatedServiceAccount:colinx,NodeName:ubuntu,HostNetwork:false,HostPID:false,HostIPC:false,SecurityContext:&PodSecurityContext{SELinuxOptions:nil,RunAsUser:nil,RunAsNonRoot:nil,SupplementalGroups:[],FSGroup:nil,},ImagePullSecrets:[{quay}],Hostname:,Subdomain:,Affinity:nil,SchedulerName:default-scheduler,InitContainers:[],AutomountServiceAccountToken:nil,Tolerations:[],HostAliases:[],PriorityClassName:,Priority:nil,},Status:PodStatus{Phase:Running,Conditions:[{Initialized True 0001-01-01 00:00:00 +0000 UTC 2018-04-12 14:19:22 +0000 UTC  } {Ready True 0001-01-01 00:00:00 +0000 UTC 2018-04-12 14:19:47 +0000 UTC  } {PodScheduled True 0001-01-01 00:00:00 +0000 UTC 2018-04-12 14:19:22 +0000 UTC  }],Message:,Reason:,HostIP:192.168.88.135,PodIP:192.168.224.6,StartTime:2018-04-12 14:19:22 +0000 UTC,ContainerStatuses:[{testapp {nil ContainerStateRunning{StartedAt:2018-04-12 14:19:23 +0000 UTC,} nil} {nil nil nil} true 0 monitor-test:latest docker://sha256:3254c818c23e92451180e9fe3aa65bf7f11fd4b39d3666f2d30eff72bb2be39c docker://59ee84d71cd3971c9aa79508b0ae33773cc969e5613cb1310f96458e21f8c16c}],QOSClass:Burstable,InitContainerStatuses:[],},}
I0412 16:27:11.151583      10 queue.go:111] syncing default/testdrain-c5589c8bf-kvlll
I0412 16:27:11.151640      10 controller.go:1180] getting endpoints for service default/ingress-default-backend and port &ServicePort{Name:,Protocol:TCP,Port:8080,TargetPort:8080,NodePort:0,}
I0412 16:27:11.151664      10 controller.go:1248] endpoints found: [{192.168.224.2 8080 0 0 false &ObjectReference{Kind:Pod,Namespace:default,Name:ingress-default-backend-6cdcc46b7c-s7l2g,UID:c8317503-3da7-11e8-b00e-000c2999b7ca,APIVersion:,ResourceVersion:126572,FieldPath:,}}]
I0412 16:27:11.151819      10 controller.go:763] creating upstream default-testdrain-80
I0412 16:27:11.151831      10 controller.go:861] obtaining port information for service default/testdrain
I0412 16:27:11.151840      10 controller.go:1180] getting endpoints for service default/testdrain and port &ServicePort{Name:http,Protocol:TCP,Port:80,TargetPort:8080,NodePort:0,}
I0412 16:27:11.151857      10 controller.go:1248] endpoints found: [{192.168.224.6 8080 0 0 false &ObjectReference{Kind:Pod,Namespace:default,Name:testdrain-c5589c8bf-kvlll,UID:7f5a0cd8-3e5c-11e8-93f0-000c2999b7ca,APIVersion:,ResourceVersion:179394,FieldPath:,}} {192.168.224.7 8080 0 0 false &ObjectReference{Kind:Pod,Namespace:default,Name:testdrain-c5589c8bf-8rrx9,UID:cff18c8d-3e5e-11e8-93f0-000c2999b7ca,APIVersion:,ResourceVersion:180152,FieldPath:,}} {192.168.224.8 8080 0 0 false &ObjectReference{Kind:Pod,Namespace:default,Name:testdrain-c5589c8bf-ml87j,UID:c83189e4-3e50-11e8-93f0-000c2999b7ca,APIVersion:,ResourceVersion:175495,FieldPath:,}}]
I0412 16:27:11.152003      10 controller.go:1072] ingress default/testdrain for host draintest.colinx.com does not contains a TLS section
I0412 16:27:11.152012      10 main.go:73] Ingress testdrain: No value found in annotation ingress.kubernetes.io/session-cookie-name. Using the default INGRESSCOOKIE
I0412 16:27:11.152020      10 main.go:83] No annotation value found in Ingress testdrain: ingress.kubernetes.io/session-cookie-strategy. Setting it to default insert
I0412 16:27:11.152026      10 main.go:91] Invalid or no annotation value found in Ingress testdrain: ingress.kubernetes.io/session-cookie-hash. Setting it to default md5
I0412 16:27:11.152738      10 main.go:73] Ingress testdrain: No value found in annotation ingress.kubernetes.io/session-cookie-name. Using the default INGRESSCOOKIE
I0412 16:27:11.152758      10 main.go:83] No annotation value found in Ingress testdrain: ingress.kubernetes.io/session-cookie-strategy. Setting it to default insert
I0412 16:27:11.152765      10 main.go:91] Invalid or no annotation value found in Ingress testdrain: ingress.kubernetes.io/session-cookie-hash. Setting it to default md5
I0412 16:27:11.152783      10 controller.go:551] replacing ingress rule default/testdrain location / upstream default-testdrain-80 (upstream-default-backend)
I0412 16:27:11.152845      10 controller.go:331] obtaining information about stream services of type TCP located in configmap 
I0412 16:27:11.152855      10 controller.go:331] obtaining information about stream services of type UDP located in configmap 
I0412 16:27:11.152870      10 controller.go:307] skipping backend reload (no changes detected)
I0412 16:27:11.162273      10 queue.go:70] queuing item &Endpoints{ObjectMeta:k8s_io_apimachinery_pkg_apis_meta_v1.ObjectMeta{Name:testdrain,GenerateName:,Namespace:default,SelfLink:/api/v1/namespaces/default/endpoints/testdrain,UID:a62af03f-3dae-11e8-b00e-000c2999b7ca,ResourceVersion:185113,Generation:0,CreationTimestamp:2018-04-11 17:34:55 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{app: testdrain,},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,},Subsets:[{[{192.168.224.7  0xc42024b540 ObjectReference{Kind:Pod,Namespace:default,Name:testdrain-c5589c8bf-8rrx9,UID:cff18c8d-3e5e-11e8-93f0-000c2999b7ca,APIVersion:,ResourceVersion:180152,FieldPath:,}} {192.168.224.8  0xc42024b550 &ObjectReference{Kind:Pod,Namespace:default,Name:testdrain-c5589c8bf-ml87j,UID:c83189e4-3e50-11e8-93f0-000c2999b7ca,APIVersion:,ResourceVersion:175495,FieldPath:,}}] [] [{http 8080 TCP}]}],}
I0412 16:27:11.162458      10 queue.go:111] syncing default/testdrain
I0412 16:27:11.500016      10 controller.go:1180] getting endpoints for service default/ingress-default-backend and port &ServicePort{Name:,Protocol:TCP,Port:8080,TargetPort:8080,NodePort:0,}
I0412 16:27:11.500116      10 controller.go:1248] endpoints found: [{192.168.224.2 8080 0 0 false &ObjectReference{Kind:Pod,Namespace:default,Name:ingress-default-backend-6cdcc46b7c-s7l2g,UID:c8317503-3da7-11e8-b00e-000c2999b7ca,APIVersion:,ResourceVersion:126572,FieldPath:,}}]
I0412 16:27:11.500306      10 controller.go:763] creating upstream default-testdrain-80
I0412 16:27:11.500351      10 controller.go:861] obtaining port information for service default/testdrain
I0412 16:27:11.500370      10 controller.go:1180] getting endpoints for service default/testdrain and port &ServicePort{Name:http,Protocol:TCP,Port:80,TargetPort:8080,NodePort:0,}
I0412 16:27:11.500492      10 controller.go:1248] endpoints found: [{192.168.224.7 8080 0 0 false &ObjectReference{Kind:Pod,Namespace:default,Name:testdrain-c5589c8bf-8rrx9,UID:cff18c8d-3e5e-11e8-93f0-000c2999b7ca,APIVersion:,ResourceVersion:180152,FieldPath:,}} {192.168.224.8 8080 0 0 false &ObjectReference{Kind:Pod,Namespace:default,Name:testdrain-c5589c8bf-ml87j,UID:c83189e4-3e50-11e8-93f0-000c2999b7ca,APIVersion:,ResourceVersion:175495,FieldPath:,}} {192.168.224.6 8080 0 0 true &ObjectReference{Kind:,Namespace:default,Name:testdrain-c5589c8bf-kvlll,UID:7f5a0cd8-3e5c-11e8-93f0-000c2999b7ca,APIVersion:,ResourceVersion:185112,FieldPath:,}}]
I0412 16:27:11.500730      10 controller.go:1072] ingress default/testdrain for host draintest.colinx.com does not contains a TLS section
I0412 16:27:11.500908      10 main.go:73] Ingress testdrain: No value found in annotation ingress.kubernetes.io/session-cookie-name. Using the default INGRESSCOOKIE
I0412 16:27:11.500935      10 main.go:83] No annotation value found in Ingress testdrain: ingress.kubernetes.io/session-cookie-strategy. Setting it to default insert
I0412 16:27:11.500943      10 main.go:91] Invalid or no annotation value found in Ingress testdrain: ingress.kubernetes.io/session-cookie-hash. Setting it to default md5
I0412 16:27:11.501208      10 main.go:73] Ingress testdrain: No value found in annotation ingress.kubernetes.io/session-cookie-name. Using the default INGRESSCOOKIE
I0412 16:27:11.501222      10 main.go:83] No annotation value found in Ingress testdrain: ingress.kubernetes.io/session-cookie-strategy. Setting it to default insert
I0412 16:27:11.501229      10 main.go:91] Invalid or no annotation value found in Ingress testdrain: ingress.kubernetes.io/session-cookie-hash. Setting it to default md5
I0412 16:27:11.501868      10 controller.go:551] replacing ingress rule default/testdrain location / upstream default-testdrain-80 (upstream-default-backend)
I0412 16:27:11.502001      10 controller.go:331] obtaining information about stream services of type TCP located in configmap 
I0412 16:27:11.502030      10 controller.go:331] obtaining information about stream services of type UDP located in configmap 
I0412 16:27:11.502043      10 controller.go:311] backend reload required
E0412 16:27:11.502352      10 runtime.go:66] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/Users/jcmoraisjr/jm/go/src/github.com/jcmoraisjr/haproxy-ingress/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:72
/Users/jcmoraisjr/jm/go/src/github.com/jcmoraisjr/haproxy-ingress/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/Users/jcmoraisjr/jm/go/src/github.com/jcmoraisjr/haproxy-ingress/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/asm_amd64.s:509
/usr/local/go/src/runtime/panic.go:491
/usr/local/go/src/runtime/panic.go:63
/usr/local/go/src/runtime/signal_unix.go:367
/Users/jcmoraisjr/jm/go/src/github.com/jcmoraisjr/haproxy-ingress/pkg/controller/stats_socket.go:132
/Users/jcmoraisjr/jm/go/src/github.com/jcmoraisjr/haproxy-ingress/pkg/controller/controller.go:173
/Users/jcmoraisjr/jm/go/src/github.com/jcmoraisjr/haproxy-ingress/pkg/common/ingress/controller/controller.go:313
/Users/jcmoraisjr/jm/go/src/github.com/jcmoraisjr/haproxy-ingress/pkg/common/ingress/controller/controller.go:169
/Users/jcmoraisjr/jm/go/src/github.com/jcmoraisjr/haproxy-ingress/pkg/common/task/queue.go:112
/Users/jcmoraisjr/jm/go/src/github.com/jcmoraisjr/haproxy-ingress/pkg/common/task/queue.go:59
/Users/jcmoraisjr/jm/go/src/github.com/jcmoraisjr/haproxy-ingress/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133
/Users/jcmoraisjr/jm/go/src/github.com/jcmoraisjr/haproxy-ingress/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134
/Users/jcmoraisjr/jm/go/src/github.com/jcmoraisjr/haproxy-ingress/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
/Users/jcmoraisjr/jm/go/src/github.com/jcmoraisjr/haproxy-ingress/pkg/common/task/queue.go:59
/usr/local/go/src/runtime/asm_amd64.s:2337
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x78 pc=0xfc7cb7]

goroutine 85 [running]:
github.com/jcmoraisjr/haproxy-ingress/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/Users/jcmoraisjr/jm/go/src/github.com/jcmoraisjr/haproxy-ingress/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x111
panic(0x10edc40, 0x1acdb40)
	/usr/local/go/src/runtime/panic.go:491 +0x283
github.com/jcmoraisjr/haproxy-ingress/pkg/controller.reconfigureBackends(0xc420150410, 0xc4201504e0, 0xc4201504e0)
	/Users/jcmoraisjr/jm/go/src/github.com/jcmoraisjr/haproxy-ingress/pkg/controller/stats_socket.go:132 +0x1b7
github.com/jcmoraisjr/haproxy-ingress/pkg/controller.(*HAProxyController).OnUpdate(0xc420294d80, 0xc4201ca880, 0x2, 0x2, 0xc4201ca890, 0x2, 0x2, 0x1b02c38, 0x0, 0x0, ...)
	/Users/jcmoraisjr/jm/go/src/github.com/jcmoraisjr/haproxy-ingress/pkg/controller/controller.go:173 +0xb6
github.com/jcmoraisjr/haproxy-ingress/pkg/common/ingress/controller.(*GenericController).syncIngress(0xc4203de240, 0x1132000, 0xc42062ad80, 0x1, 0x1)
	/Users/jcmoraisjr/jm/go/src/github.com/jcmoraisjr/haproxy-ingress/pkg/common/ingress/controller/controller.go:313 +0xb15
github.com/jcmoraisjr/haproxy-ingress/pkg/common/ingress/controller.(*GenericController).(github.com/jcmoraisjr/haproxy-ingress/pkg/common/ingress/controller.syncIngress)-fm(0x1132000, 0xc42062ad80, 0xa, 0xc4206a3d68)
	/Users/jcmoraisjr/jm/go/src/github.com/jcmoraisjr/haproxy-ingress/pkg/common/ingress/controller/controller.go:169 +0x3e
github.com/jcmoraisjr/haproxy-ingress/pkg/common/task.(*Queue).worker(0xc42035e4b0)
	/Users/jcmoraisjr/jm/go/src/github.com/jcmoraisjr/haproxy-ingress/pkg/common/task/queue.go:112 +0x35b
github.com/jcmoraisjr/haproxy-ingress/pkg/common/task.(*Queue).(github.com/jcmoraisjr/haproxy-ingress/pkg/common/task.worker)-fm()
	/Users/jcmoraisjr/jm/go/src/github.com/jcmoraisjr/haproxy-ingress/pkg/common/task/queue.go:59 +0x2a
github.com/jcmoraisjr/haproxy-ingress/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc4204617a8)
	/Users/jcmoraisjr/jm/go/src/github.com/jcmoraisjr/haproxy-ingress/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x5e
github.com/jcmoraisjr/haproxy-ingress/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc4206a3fa8, 0x3b9aca00, 0x0, 0xc420086f01, 0xc4202d2180)
	/Users/jcmoraisjr/jm/go/src/github.com/jcmoraisjr/haproxy-ingress/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xbd
github.com/jcmoraisjr/haproxy-ingress/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc4204617a8, 0x3b9aca00, 0xc4202d2180)
	/Users/jcmoraisjr/jm/go/src/github.com/jcmoraisjr/haproxy-ingress/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
github.com/jcmoraisjr/haproxy-ingress/pkg/common/task.(*Queue).Run(0xc42035e4b0, 0x3b9aca00, 0xc4202d2180)
	/Users/jcmoraisjr/jm/go/src/github.com/jcmoraisjr/haproxy-ingress/pkg/common/task/queue.go:59 +0x55
created by github.com/jcmoraisjr/haproxy-ingress/pkg/common/ingress/controller.(*GenericController).Start
	/Users/jcmoraisjr/jm/go/src/github.com/jcmoraisjr/haproxy-ingress/pkg/common/ingress/controller/controller.go:1333 +0x38f
```

**Comment by dcowden** [comment]
Ok I'm not good at Go at all, but it appears this line is the problem:

```			currentConfig.Servers[0].Locations[0].Ingress.ObjectMeta.ResourceVersion = updatedConfig.Servers[0].Locations[0].Ingress.ObjectMeta.ResourceVersion```

Since the wrapping if statement guards against the config, server, and location array lengths,one of the locations doesn't have an Ingress, or the ingress doesnt have ObjectMeta, or ObjectMetadoesnt have ResourceVersion.

**Comment by jcmoraisjr** [comment]
Couldn't reproduce this (yet). Please have a look at `quay.io/jcmoraisjr/haproxy-ingress:v0.6-2000ef7-1` - only ingress is a pointer so I just added an `&& ingress != nil`. I'll update this same issue when a proper fix is made.

**Comment by dcowden** [comment]
Ok thanks I'll try this one tomorrow. I'll also put a copy if my ingress
definition. This seems like it's probably a field you are expecting that
I've left out

**Comment by dcowden** [comment]
@jcmoraisjr I have testing results for you.  Using image `quay.io/jcmoraisjr/haproxy-ingress:v0.6-2000ef7-1`, fixed the crash-- i'm able to get a complete test. However, the test did not do as i expected.

When I drain a node manually, I see something like this:

![image](https://user-images.githubusercontent.com/1297923/38734511-ee4a4754-3ef3-11e8-810d-26ff21726083.png)

Note the drain notation on the .7 node, which i drained using this command on the haproxy-ingress container:

`echo "set server default-testdrain-80/192.168.224.7:8080 state drain " | socat /var/run/haproxy-stats.sock stdio
`
When using this approach, my tests pass. ( My tests continually loop sending requests for a list of sessions, checking that the server serving the request was the same as the original. )

When I delete a pod, v0.6-2000ef7-1 does swap things out, but it does it a different way that i don't understand. instead of drain, i see this:

![image](https://user-images.githubusercontent.com/1297923/38734734-a6bd5bd2-3ef4-11e8-9e14-4d0dc8f5b957.png)

Is this the expected behavior? In this configuration, my sticky tests fail. I'm still sorting out why, but it would be helpful to know what the difference between DRAIN without the blue, and with the blue are, and whether that's expected or not. ( I'm sorry, i know very little about haproxy)

I'll do more research also. thanks for your help on this.

**Comment by dcowden** [comment]
@jcmoraisjr also as promised here is a copy of my ingress definition. perhaps this could help determine the reason for the nil reference you did not expect:

```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: testdrain
  annotations:
    ingress.kubernetes.io/affinity : cookie
    ingress.kubernetes.io/enable-cors: "true"
    ingress.kubernetes.io/ssl-redirect : "false"
spec:
  rules:
  - host: draintest.colinx.com
    http:
      paths:
      - backend:
          serviceName: testdrain
          servicePort: 80
        path: /
```

**Comment by brianloss** [comment]
@jcmoraisjr and @dcowden, thanks for catching (and fixing) the nil pointer issue!

The way I did the drain is that the weight is set to 0 in addition to setting the drain state. When the weight is 0, HAProxy will show the line in blue. That is the only difference between the two configurations -- that the weight is 0 in one and 100 in the other. I'm not sure why HAProxy does not show the line in blue when the state is drain but the weight is 100, since the docs seem to indicate they should have the same behavior. HAProxy should be honoring the cookie and sending traffic to the terminating pod in either case.

**Comment by dcowden** [comment]
@brianloss thank you for the clarification.  My guess is that I have a working configuration from an haproxy-ingress viewpoint then-- i need to more work to figure out why my tests are not passing.  I'll do that and get back to you.

**Comment by dcowden** [comment]
hi @jcmoraisjr @brianloss I found what i was doing wrong-- the v0.6-2000ef7-1 image works correctly as far as i can tell.  I'll leave this open till the fix makes it into a regular image.  kindly let me know when that happens 

thanks!

**Comment by jcmoraisjr** [comment]
Added #138 which I hope will fix the panic. Image is `quay.io/jcmoraisjr/haproxy-ingress:v0.6-108a405`.

**Comment by dcowden** [comment]
I will retest. 
What is the timeline for v0.7?

This fix worked for us, but we need cors support before we can switch to this ingress in production.

**Comment by jcmoraisjr** [comment]
My plan is to release a new version every 3 or at most 4 months, v0.5 is the actual sad exception - it should be released a couple of months ago. v0.6 should start beta about a month from now, so perhaps we have v0.7 betas about aug/18. Otoh cors sounds to be as simple as some new headers on the response: I think it's fair to move it to v0.6 and we'll have another tester with us.

**Comment by dcowden** [comment]
awesome!   I'll be happy to test, its the least i can do. I should get this issue tested tomorrow and we can get this one closed.

**Comment by dcowden** [comment]
hi @jcmoraisjr I tested `quay.io/jcmoraisjr/haproxy-ingress:v0.6-108a405`, and it worked correctly for me.

**Comment by jcmoraisjr** [comment]
Hi, we've finally merged a lot of stuff on master and tagged the first v0.6 snapshot. Now I think I know how to test drain support myself and everything sounds to be working but another pair of eyes should help of course.

Here is the url: `quay.io/jcmoraisjr/haproxy-ingress:v0.6-snapshot.1`

Feel free to update this same issue if you have any problem - and also if you have not any problem so I'll know everything is ok.

**Comment by dcowden** [comment]
I will try to test this soon.its pretty easy, I still have my test right up.

**Comment by dcowden** [comment]
hi @jcmoraisjr i retested this, it appears to work correctly. My test plan is below:

* enable haproxy-ingress in front of three tomcats, with ingress enabled

* start three tomcats, with a java webapp that creates a cookie based session and then returns a response that contains the host name. these run behind the ingress controller

* start a script that creates sessions constantly, and also makes requests to existing sessions, checking that they come from the same backend server

* delete a pod. observe that haproxy stats screen shows the node draining, and that requests to the pod still work correctly.

 * observe that the pod is removed from haproxy stats screen when the node finishes terminating.

**Comment by dcowden** [comment]
@jcmoraisjr are you anticipating this being in the 0.6 milestone? I dont' see it in the list currently for 0.6 or 0.7

**Comment by jcmoraisjr** [comment]
It's already merged on v0.6, built on snapshot tag and published on [CHANGELOG](https://github.com/jcmoraisjr/haproxy-ingress/blob/master/CHANGELOG.md). PRs wasn't added to milestones but I'm planning to fix this ... later today.

**Comment by dcowden** [comment]
Awesome, many thanks! I'm also very appreciative to see CORS targeted for the 0.6 milestone. That'll let us switch our cluster over to this ingress controller starting with v0.6

**Comment by jcmoraisjr** [comment]
CORS is already implemented #151 should be in the next snapshot to be deployed till the next weekend.

**Comment by jcmoraisjr** [comment]
This is already implemented and apparently validated. Closing. Feel free to update this same issue if you have any problem.

**Comment by alessandroargentieri** [comment]
Hello, I've the same situation (sticky sessions with tomcat). What I don't get is:
how can ha-proxy forward the "sticky" requests to an endpoint **which is in a TERMINATING** status?
The k8s APIs will remove the endpoint **immediately**, as soon as the pod enters in a terminating status.
And also a possible **preStop** hook won't save the pod from entering immediately in a TERMINATING status, but just performs side actions while the pod is terminating (i.e. unreachable).
How 's that possible it works by using k8s APIs?

**Comment by jcmoraisjr** [comment]
hi, haproxy ingress has `drain-support` global config key that, if defined as true, the controller will add pods in the terminating status in the list of servers. The weight will be defined as 0 (zero), so new clients won't reach them.

**Comment by dcowden** [comment]
@alessandroargentieri I can confirm that this functionality works very well. We've been using tomcat with sticky sessions in production since roughly the time of this issue. It works brilliantly.   We coupled this functionality with a smart preStop that ends when there are no sessions left, so that we can turn the pods over more quickly if they didnt have sessions

**Comment by alessandroargentieri** [comment]
Now it's more clear to me. @jcmoraisjr @dcowden thank you very much for your support!
